### PR TITLE
Remove persistent tmp directory on remote hosts

### PR DESCRIPTION
**Pull request by npalaska**
Fixes issue #2085 
- Deletes persistent tool tmp directory on remote hosts when end_tools is called 
- Instead of writing persistent tool local logs on a remote file system, send them to TM via tool data sink on redis channel.

### CODE REVIEW

File: agent/util-scripts/gold/test-client-tool-meister/test-53.txt
Comment (npalaska) [reply agent/util-scripts/gold/test-client-tool-meister/test-53.txt:179]: Yeah this should really be a `debug` message, `INFO` doesn't add much value. Also this message is coming from `if-else` blocks where the `if` block of the code has an identical message with a `debug` log so I need to consistent with it

File: lib/pbench/agent/tool_meister.py
Comment (npalaska) [reply lib/pbench/agent/tool_meister.py:465]: The `threading` documentations says that the `args` is the argument tuple for the target invocation. Since its a tupple i had to insert a comma

Comment (dbutenhof) [reply lib/pbench/agent/tool_meister.py:465]: Ah; indeed. I should have thought of that. Python is weird, and sometimes ugly. ☹️

File: lib/pbench/agent/tool_meister.py
Comment (npalaska) [reply lib/pbench/agent/tool_meister.py:1563]: I agree, `self.logger.warning` would be better.

File: lib/pbench/agent/tool_meister.py
Comment (webbnh) [inline lib/pbench/agent/tool_meister.py:1544]: This seems like a bad idea:  doing it this way spreads specialize knowledge across multiple areas of the code when it could (and should) be encapsulated in one.  We should be recording the directory that we want deleted in `init_tools()` so that we _don't_ have to be making adjustments like this, here.

File: lib/pbench/agent/tool_meister.py
Comment (webbnh) [inline lib/pbench/agent/tool_meister.py:1555]: It seems like we should be complaining if there are any _files_ at **any** level of the hierarchy.  That is, we can afford to ignore subdirectories if they are empty or if they contain no subdirectories which contain any files, but we need to do a recursive file tree walk ([e.g.](https://docs.python.org/3/library/os.html#os.walk)) to find that out.

And, I wonder if, after issuing a warning, we should to go ahead and delete the files anyway.  If the files actually were valuable, then I think requiring the user to make another run is probably a lesser evil than leaving behind a bunch of useless litter (in the cases where the files are _not_ valuable).

File: lib/pbench/agent/tool_meister.py
Comment (webbnh) [inline lib/pbench/agent/tool_meister.py:1570]: Are you sure that we want to count the presence of unexpected files or the inability to delete a temporary directory as a "failure"?  They don't result in us failing to obtain the data, so they seem like things that we should just warn the user about.

**Review by webbnh — APPROVED** [review]
These changes are generally OK, but I've noted some improvements that I think you should consider.

File: lib/pbench/agent/tool_meister.py
Comment (webbnh) [inline lib/pbench/agent/tool_meister.py:427]: Just for my own edification, why is this idiom better than just
```
        for line in pipe.readline:
```
?

File: lib/pbench/agent/tool_meister.py
Comment (webbnh) [reply lib/pbench/agent/tool_meister.py:428]: Also, `line` will likely contain a `\n` at the end, do we need to strip that, or does the logger do that for us?

Comment (webbnh) [reply lib/pbench/agent/tool_meister.py:428]: Regarding my first comment, we might want to do something like:
```
    def log_subprocess_output(self, pipe):
        logger = self.logger.getChild(self.__class__.__name__)
        for line in pipe.readlines():
           logger.info(line)
```
[doc](https://docs.python.org/3/library/logging.html#logging.Logger.getChild)

**Review by ndokos — COMMENTED** [review]
LGTM. I;'ll wait until the comments are addressed.

File: lib/pbench/agent/tool_meister.py
Comment (npalaska) [reply lib/pbench/agent/tool_meister.py:428]: This is how the log message looks like without explicitly putting class name 
```
hostname.com 0016 2022-03-04 16:17:39,789 INFO 20555 139737573009152 pbench-tool-meister log_subprocess_output 428 -- b'time="2022-03-04T16:17:39Z" level=info msg=" - ipvs" source="node_exporter.go:97"\n'
```
and after putting class name
```
hostname.com 0016 2022-03-04 16:17:39,789 INFO 20555 139737573009152 pbench-tool-meister log_subprocess_output 428 -- NodeExporterTool: b'time="2022-03-04T16:17:39Z" level=info msg=" - ipvs" source="node_exporter.go:97"\n'
```
So I guess it is not adding the class name by default, but it is adding the def name and a very top level file name where it was called from (and that's the case with all of our logs at this point). Changing the logger configuration to add class name would be good but I think that should be captured in another commit. Also creating another logger just for this case would be an overkill isn't it?

File: lib/pbench/agent/tool_meister.py
Comment (npalaska) [reply lib/pbench/agent/tool_meister.py:1062]: > And, why is it using `data["directory"]` as a key? Is that value meaningful when `self._controller` is not the same as `self._hostname`?

`data["directory"]` is a value where the persistent tool data is going to be stored on the controller. The key is always (in case of persistent tools) `/var/lib/pbench-agent/<benchmark-name>/tools-<group_name>` irrespective of whether hostname is same as controller. The value for this key is either a remote tmp directory in case there is a remote or same as key in case hostname is the controller. The key just points to where the tmp tools data is stored.

File: lib/pbench/agent/tool_meister.py
Comment (npalaska) [reply lib/pbench/agent/tool_meister.py:1555]: > And, I wonder if, after issuing a warning, we should to go ahead and delete the files anyway. If the files actually were valuable, then I think requiring the user to make another run is probably a lesser evil than leaving behind a bunch of useless litter (in the cases where the files are _not_ valuable).

That's a fair point, these tmp directory is created by us and only we should have control over this directory and if we are expected to delete this directory we should delete all the files in it as well (even though we dont expect any files in there)

File: lib/pbench/agent/tool_meister.py
Comment (npalaska) [reply lib/pbench/agent/tool_meister.py]: > `iterdir()` is only going to report regular files at the first level anyway.

Yep, I should have been careful, fixed the bug in the latest commit.

File: lib/pbench/agent/tool_meister.py
Comment (npalaska) [reply lib/pbench/agent/tool_meister.py]: Fixed it

File: lib/pbench/agent/tool_meister.py
Comment (npalaska) [reply lib/pbench/agent/tool_meister.py]: Yep I concur that we should be recording the directory that we want deleted in init_tools() itself.

File: lib/pbench/agent/tool_meister.py
Comment (npalaska) [reply lib/pbench/agent/tool_meister.py]: hmm I dont know, `iter` is just something that occurred to me first I guess. But yeah I think `readlines()` is a simpler implementation to read lines.

File: lib/pbench/agent/tool_meister.py
Comment (dbutenhof) [inline lib/pbench/agent/tool_meister.py:1556]: "found [this] and will be deleted later" isn't grammatical. I also don't find the reference to "regular files" or the "will be deleted later" to be of much real value in what could turn out to be a very long message. (Also, "later" is really "immediately following this message", which means the distinction isn't very useful.)

Including the full directory path only helps in that it includes a hostname, which would otherwise be ambiguous; but I'd kinda prefer that the remote generating the message identify itself explicitly (and more compactly) instead. Listing the files, if we're implicitly deleting them, probably makes sense, but let's drop at least some of the boilerplate around the list, and put the list last.

Perhaps something like `"{hostname} unexpected temp files [{','.join(files)}]`???

Also, I think showing a separate `warning` message, from the same place, for each directory in the tree, is confusing. (Granted, we don't expect this to happen a lot.) It's only a little more work to aggregate a list that's dumped in a single message. (For bonus points, produce relative paths from the root temp directory with `Path.relative_to(*other)`.)

**Review by dbutenhof — COMMENTED** [review]
Mostly good; I'm just quibbling about the wording of the dangling file message. 😁

File: lib/pbench/agent/tool_meister.py
Comment (npalaska) [reply lib/pbench/agent/tool_meister.py:1556]: Yes I knew you would have something better to offer in terms of messaging here, so I didnt give a lot of thought to format this message :stuck_out_tongue_closed_eyes:

> Perhaps something like `"{hostname} unexpected temp files [{','.join(files)}]`???
> 
> Also, I think showing a separate `warning` message, from the same place, for each directory in the tree, is confusing. (Granted, we don't expect this to happen a lot.) It's only a little more work to aggregate a list that's dumped in a single message. (For bonus points, produce relative paths from the root temp directory with `Path.relative_to(*other)`.)

Yes I concur.

File: lib/pbench/agent/tool_meister.py
Comment (dbutenhof) [inline lib/pbench/agent/tool_meister.py:1569]: Much better, although I don't get the degenerate dict comprehension for `relative_file_path`. I also think that this list of joined lists representation is a bit weird. Since you're capturing the path of each relative to the common base directory, I would have just extended the `unexpected_files` array and done a single join at the end rather than have a list of comma-lists (assuming there's more than one directory with more than one file each, which doesn't seem to happen in the gold files I examined).

```
         unexpected_files: List[str] = []
￼        for dir, _, files in os.walk(tool_dir):
             dirpath = Path(dir).relative_to(tool_dir)
￼            if files:
￼                relative_file_path = map(
￼                    lambda x: f"{dirpath}/{x}",
￼                    files,
￼                )
￼                unexpected_files += relative_file_path
￼
￼        if unexpected_files:
￼            self.logger.warning(
￼                f"{self._hostname}: unexpected temp files {','.join(unexpected_files)}"
￼            )
```

Comment (npalaska) [reply lib/pbench/agent/tool_meister.py:1569]: Ah! Thanks I found a bug in my approach, multiple `join` inside a for loop would result in forming a single string in a list, e.g. `["/var/1,/var/2,/var/3"]` instead of `["/var/1","/var/2","/var/3"]` if there are multiple files at the same directory, and we want the latter output.

**Review by dbutenhof — APPROVED** [review]

**Review by riya-17 — APPROVED** [review]
Looks Good

File: lib/pbench/agent/tool_meister.py
Comment (portante) [inline lib/pbench/agent/tool_meister.py:1562]: Here we want to use something like:

```
self.logger.warning(
    "%s: unexpected temp files %s", self._hostname, ','.join(unexpected_files)
)
```

File: lib/pbench/agent/tool_meister.py
Comment (portante) [inline lib/pbench/agent/tool_meister.py:1568]: And again ... `"Failed to remove persistent tool data tmp directory: %s", tool_dir`.

**Review by portante — CHANGES_REQUESTED** [review]
Looks good, just need to fix the log statements.

File: lib/pbench/agent/tool_meister.py
Comment (dbutenhof) [inline lib/pbench/agent/tool_meister.py:1563]: While Peter's comment was technically valid, pragmatically we're unlikely to ever want to disable `warning` output ... and even if we did, we're still computing the `",".join` here anyway, which isn't so easy to avoid (if we really care, this would have to be transformed into an `if self.logger.isEnabledFor(logging.WARNING)` block, in which case the f-string would probably be at least as good). Still, this looks a bit more consistent than the f-string form, so that's good.

**Review by dbutenhof — APPROVED** [review]

File: lib/pbench/agent/tool_meister.py
Comment (portante) [reply lib/pbench/agent/tool_meister.py:1563]: The typical reason I make this comments is more for the discipline, since `.debug` and `.info` statements can become quite costly performance wise as a result of a lapse.

**Review by portante — APPROVED** [review]
Ship it!

File: lib/pbench/agent/tool_meister.py
Comment (ndokos) [reply lib/pbench/agent/tool_meister.py:1568]: Did you mean to change this to an error or warning? AFAICT, it's still `exception`.

**Review by ndokos — CHANGES_REQUESTED** [review]
I flagged a possible oversight.

File: lib/pbench/agent/tool_meister.py
Comment (npalaska) [reply lib/pbench/agent/tool_meister.py:1568]: I am not sure but this `else` block is no longer there, I dont know why Github did not put the `outdated` label on this conversation

Comment (npalaska) [reply lib/pbench/agent/tool_meister.py:1568]: Oh! I know how this can be confusing, this comment was made when `self.logger.exception` was under the `else` block in my first commit, but it is no longer inside the else block, Idk why githuub UI didnt detect that change here. This is what on line 1563 now `",".join(unexpected_files),`

**Review by riya-17 — APPROVED** [review]

File: lib/pbench/agent/tool_meister.py
Comment (dbutenhof) [reply lib/pbench/agent/tool_meister.py:1568]: Yeah; GitHub is attaching this conversation to the wrong place as if it were current and unchanged. That's a bit disturbing. Also, I thought I'd resolved it ... did someone unresolve it?

(Although, for the line to which it's now attached, this really *doesn't* need to be `self.logger.exception` as the exception traceback isn't particularly important and, given it would all be inside `shutil`, is more likely to be confusing than useful for debugging. *This* might as well be `self.logger.error`... or maybe even another `self.logger.warning` as we continue regardless and it doesn't affect the primary outcome of the agent activity.)

Nevertheless, I'm going to (re-)resolve this as I don't think it's critical for this PR.

**Review by ndokos — APPROVED** [review]

File: lib/pbench/agent/tool_meister.py
Comment (webbnh) [inline lib/pbench/agent/tool_meister.py:1557]: FWIW, you don't have to resort to using a `lambda` here -- `str.format()` is a callable which you can use `map()` to apply
```
                unexpected_files += map(f"{dirpath}/{{}}".format, files)
```
or you can just use a list comprehension:
```
                unexpected_files += [f"{dirpath}/{x}" for x in files]
```

File: lib/pbench/agent/tool_meister.py
Comment (webbnh) [inline lib/pbench/agent/tool_meister.py:1564]: I would use the following (but, I'm just **that** anal-retentive 😆):
```
        if unexpected_files:
            self.logger.warning(
                "%s: removing unexpected temp file%s: %s",
                self._hostname,
                "s" if len(unexpected_files) > 1 else "",
                ",".join(unexpected_files),
            )
```
(That is, pluralize `file` only if there is more than one file, and add a colon after it; also, add "removing".)

**Review by webbnh — APPROVED** [review]
Looks good to me.  Just a couple of aside comments...please mark them as resolved if you decide not to act on them.

File: lib/pbench/agent/tool_meister.py
Comment (npalaska) [reply lib/pbench/agent/tool_meister.py:1564]: I liked the suggestion but this case almost will never happen (at least we hope so) and if there are number of files getting written under this directory then either we have a bug somewhere or some other program is using the directory created by us. So I don't want to make change here after I got all the approvals :stuck_out_tongue_closed_eyes:

File: lib/pbench/agent/tool_meister.py
Comment (npalaska) [reply lib/pbench/agent/tool_meister.py:1557]: Ah! yeah I should have just used the list comprehension with f-strings. But I dont know if this change worth collapsing all the approvals.

**Pull request merged**
